addons/capetrader/capetrader.lua

```diff
@@ -0,0 +1,516 @@
+local augPaths = require('allAugPaths')
+local maxAugMap = require('maxAugMap')
+local extData = require('extdata')
+local jobToCapeMap = require('jobToCapeMap')
+
+--The following variables need to be carefully tracked and updated throughout all parts of this file.
+--TODO: Find ways to reduce the amount of global variables in this list
+local currentCape = nil
+local pathItem = nil
+local pathName = nil
+local pathIndex
+local capeHasBeenPrepared = false
+local currentlyAugmenting = false
+local firstTimeAug = false
+local timesAugmentedCount = 0
+local numberOfTimesToAugment = 0
+local tradeReady = false
+local maxAugKey = nil
+local zoneHasLoaded = true
+local inventory = nil
+
+--The following are constants.
+local dustSapThreadTradeDelay = 1
+local dyeTradeDelay = 2
+local endMessageDelay = 2
+local threadIndex = 1
+local dustIndex = 2
+local dyeIndex = 3
+local sapIndex = 4
+local inventoryBagNumber = 0
+local maxAmountDustAndThread = 20
+local maxAmountSapAndDye = 10
+local blueTextColor = 466
+local redTextColor = 123
+local greenTextColor = 158
+
+windower.register_event('addon command', function(input, ...)
+    local cmd = string.lower(input)
+    local args = {...}
+    if cmd == 'prep' then
+        if args[1] and args[2] and args[3] then
+            prepareCapeForAugments(args[2], args[1], args[3])
+        else
+            windower.add_to_chat(redTextColor, "You are missing at least one input to the prep command.")
+        end
+    elseif cmd == 'go' then
+        if zoneHasLoaded and not currentlyAugmenting then
+            if args[1] then
+                if tonumber(args[1]) then
+                    startAugmentingCape(args[1], true)
+                else
+                    windower.add_to_chat(redTextColor, 'Error: Not given a numerical argument.')
+                end
+            else
+                startAugmentingCape(1, true)
+            end
+        elseif not zoneHasLoaded then
+            windower.add_to_chat(redTextColor, 'Your inventory has not yet loaded, please try the go command again when your inventory loads.')
+        elseif currentlyAugmenting then
+            windower.add_to_chat(redTextColor, 'You are currently still augmenting a cape, please wait until the process finishes.')
+        end
+    elseif cmd == 'list' or cmd == 'l' then
+        printAugList()
+    elseif cmd == 'help' or cmd == 'h' then
+        printHelp()
+    elseif cmd == 'unload' or cmd == 'u' then
+        windower.send_command('lua unload ' .. _addon.name)
+    elseif cmd == 'reload' or cmd == 'r' then
+        windower.send_command('lua reload ' .. _addon.name)
+    else
+        windower.add_to_chat(redTextColor, 'You entered an unknown command, enter //ct help if you forget your commands.')
+    end
+end)
+
+function getItemIndex(capeOrAugItem)
+    for itemIndex, item in pairs(inventory) do
+        if item.id == capeOrAugItem.id then
+            return itemIndex
+        end
+    end
+end
+
+function getItem(itemName)
```

Comment:
This function searches through the item resources. All resource tables already provide a mechanism to do that:
```lua
    return res.items:with('name', itemName)
```

This will not match case insensitively, so the case would need to be correct. Otherwise you can do the more verbose:
```lua
    return res.items:with('name', function(name) return name:lower() == itemName end)
```

Both functions (as well as your original loop) are annoyingly inefficient. The item resources are huge with several thousand entries, and every time the function is called it searches through them all and even lowercases their names. You can avoid much of that effort by not lowercasing names unless you have to (such as when handling user input), but sometimes there's no way around it.

In this case however, you can do one of two things:
1. Memoization. That's a term describing saving function input and a hash table of results, so that only the first time the resources are searched. The second time the hashed value is taken. One possible implementation would be this:
```lua
do
    local results = {}
    getItem = function(itemName)
        itemName = itemName:lower()
        if not results[itemName] then
            results[itemName] = res.items:with('name', function(name) return name:lower() == itemName end)
        end

        return results[itemName]
    end
end
```
2. Precomputing a table of known values. This method is especially applicable if there are only a certain known number of valid values, which I think is the case here (although I'm not familiar with the augmenting mechanism, so I could be wrong). In this case you'd iterate through the resources on addon load and create a name-item mapping table. Then the `getItem` function wouldn't need to search at all, nor manage its own hash table, it would just look up the value in the global table you computed before.